File: 092/eu092.go
// project euler (projecteuler.net) problem 92
// solution by Kevin Retzke, May 2012
package main

import (
	"fmt"
)

// SquareDigits returns the sum of the squares of the digits in n.
func SquareDigits(n int) int {
	result := 0
	for ; n > 0; n = n / 10 {
		digit := n % 10
		result += digit * digit
	}
	return result
}

// NumberChain recursively calculates the end result of the number
// chain starting with n.
func NumberChain(n int) int {
	s := SquareDigits(n)
	if s == 1 || s == 89 {
		return s
	}
	return NumberChain(s)
}

func main() {
	if NumberChain(44) != 1 || NumberChain(85) != 89 {
		fmt.Println("Test Failed")
		return
	}
	result := 0
	for n := 1; n < 10000000; n++ {
		if NumberChain(n) == 89 {
			result++
		}
	}
	fmt.Println(result)
}

File: python/eu092.py
#!/usr/bin/python
# encoding: utf-8
"""
project euler (projecteuler.net) problem 92
solution by Kevin Retzke, May 2012
"""
import math

def squareDigits(n):
    digits = map(int, str(n))
    return sum([n*n for n in digits])

def numberChain(n):
    s = squareDigits(n)
    if s == 1 or s == 89:
        return s
    else:
        return numberChain(s)

if __name__ == '__main__':
    assert numberChain(44) == 1
    assert numberChain(85) == 89
    result = 0
    for n in xrange(1,10**7):
        if numberChain(n) == 89:
            result += 1
    print result
     
